### DISCUSSION TITLE
Docker container for Windows

### DISCUSSION THREAD

**Discussion by dk8pn** [DennisOSRM/hms-mqtt-publisher#81] [Ideas]
Would it be possible to setup a docker image which could be executed in "Docker for Windows"?

**Comment by DennisOSRM** [comment]
Shouldn't the amd64 container work on a Windows host just fine?

**Reply by dk8pn** [reply]
First i have to say that i am not an expert. You may be right. It's possible that i make something else wrong? This is what i get looking to details within the "Docker for Windows":

> 
2024-01-19 14:28:59 1705670939: mosquitto version 2.0.14 starting
2024-01-19 14:28:59 1705670939: Config loaded from /mosquitto/config/mosquitto.conf.
2024-01-19 14:28:59 1705670939: Starting in local only mode. Connections will only be possible from clients running on this machine.
2024-01-19 14:28:59 1705670939: Create a configuration file which defines a listener to allow remote access.
2024-01-19 14:28:59 1705670939: For more details see https://mosquitto.org/documentation/authentication-methods/
2024-01-19 14:28:59 1705670939: Opening ipv4 listen socket on port 1883.
2024-01-19 14:28:59 1705670939: Opening ipv6 listen socket on port 1883.
2024-01-19 14:28:59 1705670939: **Error: Address not available**
2024-01-19 14:28:59 1705670939: mosquitto version 2.0.14 running
> 

This information might probably also help you to identify my problem:

![image](https://github.com/DennisOSRM/hms-mqtt-publisher/assets/25267434/7ab903f7-1ed4-4fa9-a852-8714fe456c7d)

I thaught that the pathes of Linux would not fit to Windows? But this might be nonsense.

It might also be important to know that i run "Docker for Windows" on one computer and mosquitto on another computer. On the same computer where mosquitto is, i also run openhab and my final goal is to acquire the MQTT data comming from the Hoymiles inverter using openhab.

In openhab on the 2nd computer i have running a MQTT broker (port 1883). It can be configured for TCP or Websocket communication.

Is MQTT running within your docker container or is it running on my 2nd computer (sorry for such questions)? Acc. to my understanding up to now, you run a MQTT publisher (within the docker container) which sends information to my 2nd computer (ideally to my openhab MQTT broker). Last but not least i would have to setup a MQTT thing (with the MQTT broker as bridge) in order to retrieve the Hoymiles data. Hope that this is no nonsense.

I appreciate any further support very much!

**Reply by dominikandreas** [reply]
The beauty of docker is that containers run the same way no matter what the host is. So it should work fine on windows with wsl2 and docker for windows. 

Your mosquitto server log states 

>"2024-01-19 14:28:59 1705670939: Starting in local only mode. Connections will only be possible from clients running on this machine."

Maybe this is just a misconfiguration of your mosquitto broker? 

It doesn't really matter if the broker runs in docker, on the same machine or on another, as log as the 1883 port is reachable.

**Reply by dk8pn** [reply]
Thanks for this input. Mosquitto should be okay. Its a MQTT broker running in openHAB on the 2nd computer and listening on port 1883. In my view, it should not be an authentification issue. But, as i said, i am not an expert. In openHAB, i probably need to setup channels manually which are connected to the MQTT broker. But therefore i would need to know MQTT topic and payload names and syntax. Up to now, i didn't identify this information anywhere.

**Comment by dk8pn** [comment]
In openHAB i also have a so called "Home Assistant MQTT component" which could help using the "MQTT broker" as bridge.

**Comment by dk8pn** [comment]
Up to now, i have published a bunch of scrap here. I apologize for that. The major issue was that the setup using the environment variables went wrong and a had to edit the config.toml in order to get the MQTT telegrams running.

`2024-01-27 14:57:05 [2024-01-27T13:57:05Z INFO  hms_mqtt_publish] Running revision: 189413f95db615e1fad27c2e2538d1500ff446af
2024-01-27 14:57:05 [2024-01-27T13:57:05Z INFO  hms_mqtt_publish] inverter host: 192.168.1.104
2024-01-27 14:57:05 [2024-01-27T13:57:05Z INFO  hms_mqtt_publish] Publishing to simple MQTT broker
2024-01-27 14:57:05 [2024-01-27T13:57:05Z INFO  hms2mqtt::inverter] Inverter is Online` 

Used "MQTT Explorer" to verify the basic funtionality and have everything integrated now within openHAB. The only issue which i observed is that the Hoymiles cloud doesn't update as long as the container is running. But, as soon as the container is stopped, the cloud updates data, at least for a couple of hours. Have not tested longer time periods.

**Comment by DennisOSRM** [comment]
Just published a new release. The associated [GitHub action](https://github.com/DennisOSRM/hms-mqtt-publisher/actions/runs/7679710301) is running and should publish updated containers once it completes. I comes with a coop mode where you can increase the update time to 60 seconds. The cloud should then receive updates.

**Reply by dk8pn** [reply]
Many thanks for this but how can i set this time? Is an updated config.toml required?

**Reply by dk8pn** [reply]
Found your .toml file with the update interval and tested the new container. With 60 secs it still is a problem (probably because it creates a collision with the 15 min. Hoymiles interval) but with 50 secs it works out with parallel data update in the Hoymiles cloud.

**Comment by dk8pn** [comment]
Docker for Windows consumes a lot of memory. For that reason i had to use 2 computers initially. openHAB is running here on a 64bit "Intel Compute Stick" and Windows10. In order to have everything on one computer, i have tested **WSL** (Windows Subsystem for Linux) today. Despite the fact that i have zero knowledge of Linux, i have been able to install everything acc. to your github description and to edit the config.toml which was the biggest deal :-). It works out perfectly and i don't see any increase of memory consumption.
